daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -466,6 +466,42 @@ object GenTransaction extends value.CidContainer3[GenTransaction] {
           Node.GenNode.foreach3(f1, f2, f3)(node)
       }
   }
+
+  // crashes if transaction's keys contain contract Ids.
+  @throws[IllegalArgumentException]
+  def duplicatedContractKeys(tx: VersionedTransaction[NodeId, Value.ContractId]): Set[GlobalKey] = {
+
+    import GlobalKey.{assertBuild => globalKey}
+
+    case class State(active: Set[GlobalKey], duplicates: Set[GlobalKey]) {
+      def created(key: GlobalKey): State =
+        if (active(key)) copy(duplicates = duplicates + key) else copy(active = active + key)
+      def consumed(key: GlobalKey): State =
+        if (active(key)) copy(active = active - key) else this
```

Comment:
Could drop active check for this an referenced as set operation benign in excluded case.

Reply:
This check just avoids unnecessary allocation of `State` object.